### ISSUE TITLE
ETK: Improve build scripts

### ISSUE THREAD

**Issue by noahtallen** [Automattic/wp-calypso#48833]
## Overview

The [editing toolkit plugin build scripts](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/package.json) are kind of convoluted. This is difficult for new contributors having to add a significant amount of boilerplate which is the same for all submodules, but also results in very strange build output. Instead of one single JS/CSS bundle which is loaded and sent to clients, we create different assets for every single submodule (i.e. about 20 different script files.

Additionally, the final plugin artifact _includes_ the unminified source code. This means that diff creation and patching is much bigger than it needs to be.

This is probably bad for performance. However, one benefit of this approach is that we can decide to load files on a granular level. For example, since dotcom-fse is rarely used, its assets are almost never loaded. On top of that, folks have been using PHP to feature gate most of their features. For example, global styles requires theme support, and only then will its assets be loaded.

@josephscott what do you think about this issue? How do we balance not loading code on a granular level vs loading multiple files? Is it generally better to load one larger file, or multiple files which, in total, are smaller?

also cc @jsnajdr @sirreal for feedback

## Context

- n/a

**Comment by josephscott** [comment]
In general only loading what is actually needed at the time is a good thing.  @sgomes has worked with the Calypso code base more than I, he may have more specific insight for this specific case.

**Comment by scinos** [comment]
>  Is it generally better to load one larger file, or multiple files which, in total, are smaller?

In general, assuming we use HTTP/2 to serve assets and a decent client caching, it is better to serve multiple files.

**Comment by jsnajdr** [comment]
We should continue to build and deploy a separate script for each module. I don't think anyone ever suggested merging them into one -- I certainly didn't 🙂 

The issue with build scripts is that each of the scripts is built with its own NPM task, using `calypso-build --source='...'`. And that there are almost 60 NPM tasks in the `editing-toolkit/package.json` file.

The build task could be merged into one, where there is a common webpack config file with multiple entrypoints, one for each script. See how Gutenberg constructs the [`gutenbergPackages` array](https://github.com/WordPress/gutenberg/blob/master/webpack.config.js#L35-L42) and then uses it to [create an `entry` object](https://github.com/WordPress/gutenberg/blob/master/webpack.config.js#L90-L93) in the webpack config.

Adding a new script then means adding one line to a build config file.

The resulting build artifacts are exactly the same as they were before, just built with a more streamlined approach.

**Comment by sgomes** [comment]
Thanks for looping me in, @josephscott! 

To add my voice to the chorus, the end goal should always be to ship as little JS as possible, and to do so only when it's needed. It isn't exceedingly important these days whether it's coming from one or multiple scripts, as long as they all come from the same origin, so that connection setup costs are mitigated.

How to accomplish this, and how close to get to the theoretical optimal state, is a different matter, that is dependent on the code, its dependencies, how it's used, etc. For the record, though, I don't see minification and granularity as being mutually exclusive, as the issue appears to imply (sorry if I misinterpreted that); you can certainly organise things in multiple scripts that can be loaded at separate, appropriate times, while ensuring that each of these scripts is minified (and served to users over a compressed stream). That's what Calypso does, for example.

@jsnajdr's suggestion for how to organise the build is great, because it would allow you to simplify the build process and more easily apply these sorts of optimisations across the board. I'm happy to help with any questions you might have there!

**Comment by noahtallen** [comment]
> We should continue to build and deploy a separate script for each module. I don't think anyone ever suggested merging them into one -- I certainly didn't 🙂

>  I don't see minification and granularity as being mutually exclusive, as the issue appears to imply

Yes -- this was something I was curious about, and something I misunderstood. I'm glad to hear that keeping it granular will be better for performance anyways. 👍 

I'll begin by looking into Gutenberg's setup, and then seeing if I can put together a PR to apply the same to this plugin. Then I'll loop everyone in for feedback :)

**Comment by sirreal** [comment]
One important thing about a single build vs. multiple builds is that we have smarter ways of sharing assets. Right now, we share data-stores in ETK by having an independent build and registering the script it produces. Then other independent builds can depend on that as a WordPress script dependency so that it is enqueued. [That's this "common" module](https://github.com/Automattic/wp-calypso/tree/trunk/apps/editing-toolkit/editing-toolkit-plugin/common).

Within a single webpack build, we can share common assets through webpack in a variety of ways and leave more of the work to webpack.

**Comment by noahtallen** [comment]
That's really interesting. Could you point me in the direction of some resources for that?

**Comment by sirreal** [comment]
I think the [webpack docs on code splitting](https://webpack.js.org/guides/code-splitting/) is the place to start. I'm not sure what will give the best results in WordPress.

There's some prior art in Jetpack using [dynamic imports](https://webpack.js.org/guides/code-splitting/#dynamic-imports). This works well in WordPress, but requires async loading which may not be ideal for data-stores.
- [Pass the URL through to the environment.](https://github.com/Automattic/jetpack/blob/fd0a916aba87b291b491a5e6f4d44d9f88c4cf42/projects/plugins/jetpack/class.jetpack-gutenberg.php#L677-L683)
- This first thing in your script should be to [read from the environment and set the public path for webpack](https://github.com/Automattic/jetpack/blob/fd0a916aba87b291b491a5e6f4d44d9f88c4cf42/projects/plugins/jetpack/extensions/shared/public-path.js#L1-L15).
- Use [`import()` to cause webpack to split](https://github.com/Automattic/jetpack/blob/fd0a916aba87b291b491a5e6f4d44d9f88c4cf42/projects/plugins/jetpack/extensions/blocks/map/component.js#L311-L312)

Again, I don't think async is necessarily the best answer here, but may provide insight. What I don't like is [registering the script](https://github.com/Automattic/wp-calypso/blob/7f09bce3cfce10c0985e2bf4f2526673038a16e2/apps/editing-toolkit/editing-toolkit-plugin/common/index.php#L25), using [`import` as placeholder for the script dependency](https://github.com/Automattic/wp-calypso/blob/7f09bce3cfce10c0985e2bf4f2526673038a16e2/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/stores/index.ts#L4) and [even declaring it for the type system](https://github.com/Automattic/wp-calypso/blob/7f09bce3cfce10c0985e2bf4f2526673038a16e2/apps/editing-toolkit/typings/fse/index.d.ts#L1-L3), and [telling our webpack plugin how to handle that](https://github.com/Automattic/wp-calypso/blob/7f09bce3cfce10c0985e2bf4f2526673038a16e2/apps/editing-toolkit/webpack.config.js#L81-L90). This is effective, but its a very manual bridge between the two systems whereas the Jetpack solution feels like an approach that works well within the systems of webpack and WordPress script dependencies. With a single build, I think we should be able to get something better.

This isn't the first thing to tackle in these changes 😅  It's working fine now. After we work towards a single build this type of change should be much simpler to take on.

**Comment by noahtallen** [comment]
Very interesting, I'll research those more. I'm definitely not very familiar with that. 

It does sound like a single build will be an excellent thing to implement next, so I'll try to do that. It doesn't sound too hard ™️
